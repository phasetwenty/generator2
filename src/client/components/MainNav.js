/**
 * Copyright 2017 Christopher Haverman
 * All Rights Reserved
 **/
import {Link} from 'react-router-dom';
import {Nav, NavItem} from 'reactstrap';
import PropTypes from 'prop-types';
import React from 'react';

const MainNav = (props) => {
  return (
    <Nav vertical>
      {
        props.items.map((item, index) => {
          return (
            <div key={index}>
              {renderItem(item, index, 1)}
              {item.subcategories.map((item, index) => renderItem(item, index, 2))}
            </div>
          );
        })
      }
    </Nav>
  );
};

MainNav.propTypes = {
  items: PropTypes.array.isRequired
};

function renderItem(item, key, depth) {
  const offsetClass = `ml-${depth}`;
  const itemOffsetClass = `ml-${depth + 1}`;
  return (
    <div key={key}>
      <MainNavSection key={key} offsetClass={offsetClass} text={item.name}/>
      {
        item.objects.map((pair, index) => {
          return <MainNavItem key={index} offsetClass={itemOffsetClass} target={`/${pair[1]}`} text={pair[0]}/>;
        })
      }
    </div>
  );
}


const MainNavSection = (props) => {
  return <NavItem className={props.offsetClass}>{props.text}</NavItem>;
};

MainNavSection.propTypes = {
  offsetClass: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired
};

const MainNavItem = (props) => {
  const {offsetClass, target, text} = props;
  return <NavItem className={offsetClass}><Link to={target}>{text}</Link></NavItem>;
};

MainNavItem.propTypes = {
  offsetClass: PropTypes.string.isRequired,
  target: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired
};

export default MainNav;
