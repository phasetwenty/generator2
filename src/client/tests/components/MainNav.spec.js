/**
 * Copyright 2017 Christopher Haverman
 * All Rights Reserved
 **/
/* global beforeEach, describe, expect, test */
import MainNav, {MainNavItem, MainNavSection} from '../../components/MainNav';
import {MemoryRouter} from 'react-router-dom';
import React from 'react';
import renderer from 'react-test-renderer';

describe('MainNav', () => {
  const items = [
    {
      "name": "Top-level",
      "subcategories": [{"name": "Second-level", "objects": [["Thieves"], ["Nobles"]]}],
      "objects": []
    }
  ];

  test('matches snapshot', () => {
    const currentSnapshot = renderer.create(
        <MemoryRouter>
          <MainNav items={items}/>
        </MemoryRouter>
    ).toJSON();
    expect(currentSnapshot).toMatchSnapshot();
  });
});
