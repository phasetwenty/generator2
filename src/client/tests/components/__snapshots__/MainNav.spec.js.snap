// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNav matches snapshot 1`] = `
<ul
  className="nav flex-column"
>
  <div>
    <div>
      <li
        className="ml-1 nav-item"
      >
        Top-level
      </li>
    </div>
    <div>
      <li
        className="ml-2 nav-item"
      >
        Second-level
      </li>
      <li
        className="ml-3 nav-item"
      >
        <a
          href="/undefined"
          onClick={[Function]}
        >
          Thieves
        </a>
      </li>
      <li
        className="ml-3 nav-item"
      >
        <a
          href="/undefined"
          onClick={[Function]}
        >
          Nobles
        </a>
      </li>
    </div>
  </div>
</ul>
`;
